tools/deadlock_detector.py

```diff
@@ -0,0 +1,282 @@
+            [ stack trace ]
+
+        Mutex m previously acquired by thread T here:
+            [ stack trace ]
+
+    for T in all threads:
+        Thread T was created here:
+            [ stack trace ]
+    '''
+    edges = find_cycle(graph)
+    if not edges:
+        return False
+
+    # List of mutexes in the cycle, first and last repeated
+    nodes_in_order = []
+    # Map mutex address -> readable alias
+    node_addr_to_name = {}
+    for counter, (m, n) in enumerate(edges):
+        nodes_in_order.append(m)
+        node_addr_to_name[m] = 'Mutex M%d (0x%016x)' % (counter, m)
+    nodes_in_order.append(nodes_in_order[0])
+
+    print('----------------\nPotential Deadlock Detected!\n')
+    print(
+        'Cycle in lock order graph: %s\n' %
+        (' => '.join([node_addr_to_name[n] for n in nodes_in_order]))
+    )
+
+    # Set of threads involved in the lock inversion
+    thread_pids = set()
+
+    # For each edge in the cycle, print where the two mutexes were held
+    for (m, n) in edges:
+        thread_pid = graph[m][n]['thread_pid']
+        thread_comm = graph[m][n]['thread_comm']
+        first_mutex_stack_id = graph[m][n]['first_mutex_stack_id']
+        second_mutex_stack_id = graph[m][n]['second_mutex_stack_id']
+        thread_pids.add(thread_pid)
+        print(
+            '%s acquired here while holding %s in Thread %d (%s):' % (
+                node_addr_to_name[n], node_addr_to_name[m], thread_pid,
+                thread_comm
+            )
+        )
+        print_stack_trace_fn(second_mutex_stack_id)
+        print('')
+        print(
+            '%s previously acquired by the same Thread %d (%s) here:' %
+            (node_addr_to_name[m], thread_pid, thread_comm)
+        )
+        print_stack_trace_fn(first_mutex_stack_id)
+        print('')
+
+    # Print where the threads were created, if available
+    for thread_pid in thread_pids:
+        parent_pid, stack_id, parent_comm = thread_to_parent.get(
+            thread_pid, (None, None, None)
+        )
+        if parent_pid:
+            print(
+                'Thread %d created by Thread %d (%s) here: ' %
+                (thread_pid, parent_pid, parent_comm)
+            )
+            print_stack_trace_fn(stack_id)
+        else:
+            print(
+                'Could not find stack trace where Thread %d was created' %
+                thread_pid
+            )
+        print('')
+    return True
+
+
+def strlist(s):
+    '''
+    Given a comma-separated string, returns a list of substrings
+    '''
+    return s.strip().split(',')
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=(
+            'Detect potential deadlocks (lock inversions) in a running binary.'
+            ' Must be run as root.'
+        )
+    )
+    parser.add_argument('binary', type=str, help='Absolute path to binary')
+    parser.add_argument('pid', type=int, help='Pid to trace')
```

Comment:
We can resolve the pid to a binary by looking at /proc/pid/exe, but we had issues with this because:

- Some of our binaries have `:` in the path name (e.g. IP addresses)
- On kernels without this recent kernel patch (https://lkml.org/lkml/2017/1/13/585, submitted 2 weeks ago), we can't attach a uprobe on that binary because of the `:` in the path name
- However, if we have a symlink to the binary, it should work.